[QOLSVC-10243] exclude dataset search results from bot indexing

ThrawnCA opened the pull request:
Add a `meta` tag to exclude indexing (but allow following links to sub-pages) any time there are dataset search parameters, #9063

ThrawnCA committed:
[QOLSVC-10243] exclude dataset search results from bot indexing

duttonw reviewed (approved):
would this include 'tags' 'groups' 'organisation' filters?

why not just block the 'search' page from being indexed full stop?

ThrawnCA commented:
> would this include 'tags' 'groups' 'organisation' filters?

Yes, they are part of the query string, thus they populate `request.args`, and so they trigger it.

> why not just block the 'search' page from being indexed full stop?

Because we do want Google to see `/dataset`. We just don't want an extra copy of it for every permutation of search parameters.

duttonw commented:
ok lets get this passing cicd and then create a PR to upstream as this would be useful (may need a ckan config flag for good measure)

ThrawnCA committed:
[QOLSVC-10243] exclude group search results from bot indexing

- Move the robots meta tag into the base template, controlled by a flag that most views won't set

duttonw commented on ckan/views/group.py:
```diff
@@ -130,6 +130,7 @@ def index(group_type: str, is_organization: bool) -> str:
 
     extra_vars["q"] = q
     extra_vars["sort_by_selected"] = sort_by
+    extra_vars['has_search_args'] = True if request.args else False
```
does this also cover organizations?

ThrawnCA replied:
Yes. Line 194 selects the template name based on the type, group vs org.

EricSoroos commented on ckan/templates/base.html:
```diff
@@ -26,6 +26,10 @@
       <meta name="{{ g.csrf_field_name }}" content="{{ csrf_token() }}" />
+      {# Search result pages shouldn't be indexed by bots. #}
+      {% if has_search_args %}
+        <meta name="robots" content="noindex, follow" />
```
I think page=? is a legit argument here, allowing for paging through the basic result set.

duttonw replied:
Hi @EricSoroos ,

Reviewing the docs, they will still page through all pages and permutations it just won't be captured and displayed to the user. i.e. do you want the real dataset buried under search page permutation for every tag, solr key word etc?

https://developers.google.com/search/docs/crawling-indexing/robots-meta-tag
https://docs.squiz.net/matrix/version/latest/developing/search/prepare-your-site-for-search/robots-txt.html#robots-meta-tags-control-web-robots-at-the-page-level
https://www.bing.com/webmasters/help/which-robots-metatags-does-bing-support-5198d240
https://seranking.com/blog/guide-meta-tag-robots-x-robots-tag/

ThrawnCA replied:
Search results can change, meaning that the content of a page of results can change. Therefore we don't want eg Google caching them.

Sending people to the base search page is fine. Sending them to specific datasets is fine. Sending them to "page 2 of the list of datasets" is too volatile.

That said, we wouldn't want to forbid bots from _visiting_ every page of results. We just don't want them to _index_ every page. Thus "noindex, follow".
